### PR TITLE
Cit solver optimalization

### PR DESCRIPTION
**Pull request by richtja**
Fixes and optimizations of cit_plugin constraint solver.

### CODE REVIEW

File: optional_plugins/varianter_cit/avocado_varianter_cit/Cit.py [hunk 1]
```diff
@@ -303,15 +303,15 @@ def compute_hamming_distance(self, row):
         return distance
 
     def create_random_row_with_constraints(self):
-        row = []
-        data_matrix = []
-        for parameter in self.data:
-            data_matrix.append(list(range(0, parameter)))
+        """
```
Comment (beraldoleal) [inline optional_plugins/varianter_cit/avocado_varianter_cit/Cit.py:306]: We should use PEP 257 for docstrings.

File: optional_plugins/varianter_cit/avocado_varianter_cit/Cit.py [hunk 2]
```diff
@@ -303,15 +303,15 @@ def compute_hamming_distance(self, row):
...
+        Creates a new test case random row and the row meets the constraints
+        :return: new random row
+        :rtype: list
+        """
+        row = [-1]*len(self.data)
 
-        # delete the forbidden values ​by constraints
-        self.solver.clean_data_matrix(data_matrix)
-        for parameter, possible_values in enumerate(data_matrix):
+        for parameter in random.sample(range(len(self.data)), len(self.data)):
```
Comment (beraldoleal) [inline optional_plugins/varianter_cit/avocado_varianter_cit/Cit.py:313]: You are calculating the `len(data)` 3 times here. Maybe we could add a `data_size = len(data)` and use this.

File: optional_plugins/varianter_cit/avocado_varianter_cit/Parameter.py [hunk 3]
```diff
@@ -20,34 +20,25 @@ def __hash__(self):
...
-                self.constrained_values.add(pair.value)
-                value = pair.value
-            else:
-                array.append(pair)
-        if len(self.constrained_values) == len(self.values):
-            self.is_full = True
+        self._constrained_values_number = 0
+        self.constraints = [None]*len(values)
+
+    def add_constraint(self, constraint, value, index):
```
Comment (beraldoleal) [inline optional_plugins/varianter_cit/avocado_varianter_cit/Parameter.py]: missing docstring.

File: optional_plugins/varianter_cit/avocado_varianter_cit/Solver.py [hunk 4]
```diff
@@ -107,18 +110,36 @@ def clean_hash_table(self, combination_matrix, t_value):
...
-            for constraint in parameter_constraint:
-                constraint = constraint[0]
-                try:
-                    data_matrix[constraint.name].remove(constraint.value)
-                except ValueError:
-                    # this value was already deleted
-                    pass
+    def get_possible_values(self, row, parameter):
+        """
+        Computes all possible values for the given parameter based on
```
Comment (beraldoleal) [inline optional_plugins/varianter_cit/avocado_varianter_cit/Solver.py:135]: Docstring is not following the PEP 257.

File: optional_plugins/varianter_cit/avocado_varianter_cit/Solver.py [hunk 5]
```diff
@@ -107,18 +110,36 @@ def clean_hash_table(self, combination_matrix, t_value):
...
+                if not is_ok:
+                    return False
+            return True
+
+        possible_values = []
+        parameter = self.parameters[parameter]
+        for value, constraints in enumerate(parameter.constraints):
+            if constraints is None or is_permitted_value(constraints):
+                possible_values.append(value)
+        return possible_values
```
Comment (beraldoleal) [inline optional_plugins/varianter_cit/avocado_varianter_cit/Solver.py:165]: Just a suggestion, maybe we could replace this code by a `filter`

File: optional_plugins/varianter_cit/avocado_varianter_cit/Solver.py [hunk 6]
```diff
@@ -1,6 +1,28 @@
 import itertools
 
-from .Parameter import Parameter
+
+class Parameter:
```
Comment (beraldoleal) [inline optional_plugins/varianter_cit/avocado_varianter_cit/Solver.py:4]: missing docstring.

File: optional_plugins/varianter_cit/avocado_varianter_cit/Solver.py [hunk 7]
```diff
@@ -1,6 +1,28 @@
...
+
+    def __init__(self, name, parameter_id, values):
+        self.name = name
+        self.id = parameter_id
+        self.values = values
+        self._constrained_values_number = 0
+        self.constraints = [None]*len(values)
+
+    def add_constraint(self, constraint, value, index):
```
Comment (beraldoleal) [inline optional_plugins/varianter_cit/avocado_varianter_cit/Solver.py:13]: Missing docstring.

File: optional_plugins/varianter_cit/avocado_varianter_cit/Cit.py (hunk 2 above)
Comment (richtja) [reply optional_plugins/varianter_cit/avocado_varianter_cit/Cit.py:313]: Yes, you are right the `data_size` will be better.

File: optional_plugins/varianter_cit/avocado_varianter_cit/Solver.py (hunk 5 above)
Comment (richtja) [reply optional_plugins/varianter_cit/avocado_varianter_cit/Solver.py:165]: That is a great idea, I will use that thanks.

**Comment by richtja** [comment]
@beraldoleal thanks for the comments, I'll use them and also do the style fixes.

**Comment by richtja** [comment]
Closing in favour of #3755

**Pull request closed without merging**